### DISCUSSION TITLE
Buildpack to C and C++ Framewwork

### DISCUSSION THREAD

**Discussion by karlos84-sudo** [paketo-buildpacks/feedback#207] [Q&A]
Our team is finding to support some applications done on C and C++ frameworks,  but I don't see any buildpack for these on Paketo repositories. Please I would like to get advice/help to build my own buildpack to these frameworks.
Thanks in advance.

**Comment by dmikusa** [comment]
You're correct that there are not presently any Paketo C/C++ buildpacks. Our builders do ship a lot of the tools that you'd need to compile C/C++ code though because other buildpacks like Python/Ruby/Node.js use those tools to compile native extensions.

I think the primary issue with a Paketo C/C++ buildpack is knowing what to make it do. It's easy enough to detect C/C++ code, but how you build that code can vary wildly. You've got GNU build tools, CMake, custom Makefiles, and other build systems.

You might want to give making a buildpack for your needs a try. There's a tutorial here, https://buildpacks.io/docs/buildpack-author-guide/create-buildpack/, which explains the basics of making a buildpack with some shell script. That's a very approachable way of making the buildpack. If you have a limited scope of tools, you might be able to get something working without too much fuss.

The other thing you could potentially do is to compile your C/C++ application outside of buildpacks, like in your CI system, and then use the Procfile buildpack to containerize the binary.

If your CI builds the app and makes a folder with the binary and a Procfile, you can then `pack build my-c-app`. It should just run the procfile buildpack and set up the container to run using the commands in your Procfile. You could very likely use the `paketobuildpacks/builder-jammy-tiny` which is small, like 14M, or possibly even the static builder if your binary is fully static. It all just depends on what libraries need to existing in the runtime image for the app to execute properly.

Hope that helps!

**Comment by robdimsdale** [comment]
Buildpacks can be as simple as a couple of small shell scripts.

I'd start with the Cloud Native Buildpack docs: https://buildpacks.io/docs/buildpack-author-guide/create-buildpack/

The Paketo buildpacks leverage shared libraries like [libpak](https://github.com/paketo-buildpacks/libpak) and [packit](https://github.com/paketo-buildpacks/packit) because there's a lot of common functionality across buildpacks (like logging, layer management, etc).
